### DISCUSSION TITLE
Self-assessment pages for high school math students!

### DISCUSSION THREAD

**Discussion by axiomtutor** [pyscript/pyscript#1644] [Show and tell]
I want to say thank you to the developers here -- I have been using PyScript in combo with Quarto to make what I think are quite attractive and useful websites so that high school students can practice their math lessons.  I use REPLs and some backend Python (sometimes with the help of SymPy) to allow them to automate certain steps and focus on the conceptual lessons.

https://axiomtutor.quarto.pub/natalies-quarto/

In the near-ish future I have some thoughts about how I could make an entire web textbook in this fashion!

**Comment by fpliger** [comment]
@axiomtutor this is really cool! Education and this type of flow is definitely one of the primary use cases we had in mind when creating and developing PyScript and this is a great validation.

Would love to hear more about your thoughts!

**Reply by axiomtutor** [reply]
Sure, if you have any specific questions let me know!  

But where I am right now with this is: I want to be able to incorporate interactive three-dimensional graphics.  In particular, I plan on eventually building a multivariable calculus textbook in this style.  That's meant taking a particularly painful detour down learning how to use Three.js.  But just within the last few minutes I've had a bit of a break through, so hopefully I'll soon be done with Three.js and will get back to the stuff I'm building in PyScript.

A fantasy list of features that I would build (fantasy because it would probably take me so long to implement them all) would include 

* Automatically, randomly generated exercise problems.
* Automatic feedback on answers to exercises.
* Sequential hint-giving (Click a button, get a hint.  Still can't figure it out?  Click another button, get more hints.).
* Use a REPL with SymPy and matplotlib (presumably with a layer of my own code, so that students don't have spend too much time learning how to program in a math class), to replace "graphing calculators" that high schools are filled with.  

Anyway, I'm not sure if those kinds of thoughts were what you were asking about.  They're mostly things I'm thinking of implementing through PyScript.  I would imagine these use cases would be a little too narrow for PyScript to build -- but if guys ever get bored, please feel free to!  ;)

One thing that I am eventually going to struggle with -- which you guys might be interested in thinking about too, but I dunno -- is putting LaTeX in graphical elements, like a Three.js animation.  As I currently understand it, the way to do that is to have MathJax render the LaTeX source, then have Three.js scrape the rendered SVG out of the page (say in a hidden element) and put it into the canvas.  This sounds at least complicated and at worst so fragile that it could break.  I don't know, I haven't tried it yet.  But that future struggle has been on my mind.

Oh also, you might want to know about one small-ish struggle I had with PyScript.  I made a matplotlib graph of a function.  I then used PyScript to add a mouseover listener.  I extracted the x-coordinate of the mouse position, and used it to plot the point on the function with the given x-coordinate.  

One struggle I had was knowing how to get that x-coordinate.  Eventually I learned about Python's `dir` function to find out all the methods that an event object supports.  That's how I found out about the `offsetX` method for events.   I couldn't really find any information in PyScript's documentation about event objects.  There's this page: https://docs.pyscript.net/latest/reference/API/attr_to_event.html  But that seems to be for elements declared in HTML rather than generated by proxies.  

Also, `offsetX` is the closest thing to what I wanted.  But I still had to guess-and-check numbers for a shift and scale so that the mouse coordinates would correspond to graph coordinates.  I'm not sure if there is any other way to do this.  But if you guys have or ever make a better way of getting local coordinates inside a matplotlib rendering, that could be a help!

Anyway, this got rather long and rambly, hope it's not obnoxious!  Thank you guys for the awesome product!
